hugr-core/src/package.rs

```diff
@@ -1,41 +1,43 @@
 //! Bundles of hugr modules along with the extension required to load them.
 
 use derive_more::{Display, Error, From};
-use std::collections::HashMap;
+use itertools::Itertools;
 use std::path::Path;
-use std::sync::Arc;
 use std::{fs, io, mem};
 
 use crate::builder::{Container, Dataflow, DataflowSubContainer, ModuleBuilder};
-use crate::extension::{ExtensionRegistry, ExtensionRegistryError};
+use crate::extension::resolution::ExtensionResolutionError;
+use crate::extension::{ExtensionId, ExtensionRegistry};
 use crate::hugr::internal::HugrMutInternals;
-use crate::hugr::{HugrView, ValidationError};
+use crate::hugr::{ExtensionError, HugrView, ValidationError};
 use crate::ops::{FuncDefn, Module, NamedOp, OpTag, OpTrait, OpType};
-use crate::{Extension, Hugr};
+use crate::Hugr;
 
-#[derive(Debug, Default, Clone, serde::Serialize, serde::Deserialize)]
-/// Package of module HUGRs and extensions.
-/// The HUGRs are validated against the extensions.
+#[derive(Debug, Default, Clone, PartialEq, serde::Serialize, serde::Deserialize)]
+/// Package of module HUGRs.
 pub struct Package {
     /// Module HUGRs included in the package.
     pub modules: Vec<Hugr>,
-    /// Extensions to validate against.
-    pub extensions: Vec<Arc<Extension>>,
+    /// Extensions used in the modules.
+    ///
+    /// This is a superset of the extensions used in the modules.
+    pub extensions: ExtensionRegistry,
```

Comment:
Good question!

This PR does **not** change the serialization format, it's purely change to the runtime representation of hugrs.

In our current serialization model, packages can embed arbitrary extension definitions in the json.
Since the modules do not necessarily use those, if we didn't have this field we would lose track of them.

We could update the serialization and (optionally) include modules in the serialized hugr, but I think we should rather wait for `hugr-model` first.